crates/pglt_lexer/src/diagnostics.rs

```diff
@@ -0,0 +1,63 @@
+use pglt_diagnostics::{Diagnostic, MessageAndDescription};
+use text_size::TextRange;
+
+/// A specialized diagnostic for scan errors.
+///
+/// Scan diagnostics are always **fatal errors**.
+#[derive(Clone, Debug, Diagnostic, PartialEq)]
+#[diagnostic(category = "syntax", severity = Fatal)]
+pub struct ScanError {
+    /// The location where the error is occurred
+    #[location(span)]
+    span: Option<TextRange>,
+    #[message]
+    #[description]
+    pub message: MessageAndDescription,
+}
+
+impl ScanError {
+    pub fn from_pg_query_err(err: pg_query::Error, input: &str) -> Vec<Self> {
+        let err_msg = err.to_string();
+        let re = regex::Regex::new(r#"at or near "(.*?)""#).unwrap();
+        let mut diagnostics = Vec::new();
+
+        for captures in re.captures_iter(&err_msg) {
+            if let Some(matched) = captures.get(1) {
+                let search_term = matched.as_str();
+                for (idx, _) in input.match_indices(search_term) {
+                    let from = idx;
+                    let to = from + search_term.len();
+                    diagnostics.push(ScanError {
+                        span: Some(TextRange::new(
+                            from.try_into().unwrap(),
+                            to.try_into().unwrap(),
+                        )),
+                        message: MessageAndDescription::from(err_msg.clone()),
+                    });
+                }
+            }
```

Comment:
we try to find the problematic text in the input to put a range on the diagnostics. this is not safe in any way but I find it nicer to show too many diagnostics where one is for sure valid than none and have the user look through the text.